File: src/test/java/cn/sd/jrz/jtool/asynchronous/AsyncTest.java
package cn.sd.jrz.jtool.asynchronous;

import cn.sd.jrz.jtool.exception.Catch;
import org.junit.Test;

/**
 * @author 江荣展
 * Date: 2019-04-12
 * Time: 18:08
 */
public class AsyncTest {

    @Test
    public void runTest() {
        Async.run(() -> {
            System.out.println("1");
            Thread.sleep(2000);
            System.out.println("2");
        });

        Catch.run(() -> Thread.sleep(3 * 1000));
    }

    @Test
    public void getTest() {
        JBox<String> box = Async.get(() -> {
            System.out.println("1");
            Thread.sleep(2000);
            return "123";
        });
        String result = box.await(1);
        System.out.println(result);
    }

    @Test
    public void getTest2() {
        JBox<String> box = Async.get(() -> {
            System.out.println("1");
            Thread.sleep(2000);
            return "123";
        });
        String result = box.await(3);
        System.out.println(result);
    }

    @Test
    public void getTest3() {
        JBox<String> box = Async.get(() -> {
            System.out.println("1");
            return "123";
        });
        Catch.run(() -> Thread.sleep(2 * 1000));
        String result = box.await(3);
        System.out.println(result);
    }

    @Test
    public void getTest4() {
        JBox<String> box = Async.get(() -> {
            System.out.println("1");
            return "123";
        });
        Catch.run(() -> Thread.sleep(2 * 1000));
        String result = box.await();
        System.out.println(result);
    }
}

File: src/test/java/cn/sd/jrz/jtool/asynchronous/AsyncTest.kt
package cn.sd.jrz.jtool.asynchronous

import kotlinx.coroutines.delay
import org.junit.Test

/**
 * @author 江荣展
 * @date 2020/7/1
 */

class AsyncTest {
    @Test
    fun runTest() {
        Async.running {
            println("1")
            delay(2000)
            println("2")
        }
        Thread.sleep(3 * 1000)
    }

    @Test
    fun getTest() {
        Async.timeoutGet(1) {
            println("1")
            delay(2000)
            "123"
        }?.let { println(it) }
    }

    @Test
    fun getTest2() {
        Async.timeoutGet(3) {
            println("1")
            delay(2000)
            "123"
        }?.let { println(it) }
    }

    @Test
    fun getTest3() {
        val result = Async.waitGet {
            println("1")
            delay(2000)
            "123"
        }
        println(result.await(1))
    }

    @Test
    fun getTest4() {
        val result = Async.waitGet {
            println("1")
            delay(2000)
            "123"
        }
        result.value
        println(result.await(3))
    }
}